### PR TITLE
Pegasus

### PR DESCRIPTION
**Pull request by dafrenchyman**
- Added a new image that uses [pegasus](https://pegasus-frontend.org) as it's front-end
- Added Cemu
- Added Dolphin
- `es-de` & `pegasus` images use `sway` as the window manager if `GAMESCOPE` not selected.

### CODE REVIEW

**Comment by dafrenchyman** [comment]
Corrected the errors mentioned in the previous Draft PR: https://github.com/games-on-whales/gow/pull/163

File: images/pegasus/scripts/startup.sh [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  export GAMESCOPE_REFRESH=${GAMESCOPE_REFRESH:-60}
+  export GAMESCOPE_MODE=${GAMESCOPE_MODE:-"-b"}
+  /usr/games/gamescope "${GAMESCOPE_MODE}" -W "${GAMESCOPE_WIDTH}" -H "${GAMESCOPE_HEIGHT}" -r "${GAMESCOPE_REFRESH}" -- pegasus-fe
+else
+  echo "Sway - Starting Pegasus"
+  export XDG_SESSION_TYPE=wayland
+  mkdir -p $HOME/.config/sway/
+  cp /cfg/sway/config $HOME/.config/sway/config
+  exec sway --unsupported-gpu
+fi
```
Comment (ABeltramo) [inline images/pegasus/scripts/startup.sh:136]: This should be moved in a separate script like `launch-comp.sh` possibly in the `base-app` image and just referenced in all images; we shouldn't copy paste it around like this.

File: images/pegasus/configs/sway/config
Comment (ABeltramo) [inline images/pegasus/configs/sway/config:1]: This should be moved to `base-app`

File: images/pegasus/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    libxcb-xkb1 \
+    libxkbcommon-x11-0 \
+    libzstd1 \
+    policykit-1 \
+    unzip && \
+    wget $PEGASUS_DOWNLOAD --output-document=pegasus-frontend.deb && \
+    dpkg -i pegasus-frontend.deb
+
+# Install sway
+RUN apt-get install -y sway xwayland
```
Comment (ABeltramo) [inline images/pegasus/Dockerfile:146]: This should be moved to `base-app`

**Review by ABeltramo — CHANGES_REQUESTED** [review]
Thanks for all your work on this, it's highly appreciated!
Apart from the few changes that I've commented I think that there shouldn't be no need to manually set all emulators joypad settings. I'll try it out on the latest Wolf version to see if the new changes will allow to avoid that and I'll let you know how it goes..

File: images/pegasus/scripts/startup.sh [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+gow_log "CEMU - Setting default sound device and run in fullscreen, if not edited"
+cp -u /cfg/cemu/settings.xml /cfg/cemu/settings_new.xml
+searchString="<TVDevice>replace_me</TVDevice>"
+replaceString="<TVDevice>${PULSE_SINK}</TVDevice>"
+sed -i -e "s|$searchString|$replaceString|g" /cfg/cemu/settings_new.xml
+searchString="<PadDevice>replace_me</PadDevice>"
+replaceString="<PadDevice>${PULSE_SINK}</PadDevice>"
+sed -i -e "s|$searchString|$replaceString|g" /cfg/cemu/settings_new.xml
+# Set the date of the file back to the original date
+touch -r /cfg/cemu/settings.xml /cfg/cemu/settings_new.xml
```
Comment (ABeltramo) [inline images/pegasus/scripts/startup.sh:89]: I get a permission denied on my side when doing this, `/cfg/` has been created by `root` in the Dockerfile; in the startup file we should only modify `$HOME`

Comment (ABeltramo) [inline images/pegasus/scripts/startup.sh:89]: Maybe this should become something like:
```bash
CEMU_CFG_DIR=$HOME/.config/Cemu

mkdir -p $CEMU_CFG_DIR
gow_log "CEMU - Setting default sound device and run in fullscreen, if not edited"
# Set the date of the file back to the original date
#touch -r /cfg/cemu/settings.xml /cfg/cemu/settings_new.xml
cp -u /cfg/cemu/settings.xml $CEMU_CFG_DIR/settings.xml
searchString="<TVDevice>replace_me</TVDevice>"
replaceString="<TVDevice>${PULSE_SINK}</TVDevice>"
sed -i -e "s|$searchString|$replaceString|g" $CEMU_CFG_DIR/settings.xml
searchString="<PadDevice>replace_me</PadDevice>"
replaceString="<PadDevice>${PULSE_SINK}</PadDevice>"
sed -i -e "s|$searchString|$replaceString|g" $CEMU_CFG_DIR/settings.xml
```
and only do it if the file is missing?
```bash
if test -f $CEMU_CFG_DIR/settings.xml
```

File: images/pegasus/scripts/startup.sh (hunk 1 above)
Comment (dafrenchyman) [reply images/pegasus/scripts/startup.sh]: Done!

File: images/pegasus/configs/sway/config
Comment (dafrenchyman) [reply images/pegasus/configs/sway/config:1]: Done!

File: images/pegasus/Dockerfile (hunk 2 above)
Comment (dafrenchyman) [reply images/pegasus/Dockerfile]: Done!

File: images/pegasus/scripts/startup.sh (hunk 3 above)
Comment (dafrenchyman) [reply images/pegasus/scripts/startup.sh]: Fixed it!

Comment (dafrenchyman) [reply images/pegasus/scripts/startup.sh]: I added this idea to it. Much better then doing the whole timestamp checking

**Comment by dafrenchyman** [comment]
All comments addressed. There is one issue though. I still can't get the latest `master` to run retroarch, yuzu, and a few other emulators for some reason. I tested this "blindly".

File: images/base-app/scripts/init-gamescope.sh
Comment (dafrenchyman) [inline images/base-app/scripts/init-gamescope.sh:1]: I made these bash scripts executable. I didn't delete them. Github is doing something weird to them.

File: images/base-app/scripts/startup.sh
Comment (dafrenchyman) [inline images/base-app/scripts/startup.sh:1]: Executable

File: images/base-app/scripts/wait-x11
Comment (dafrenchyman) [inline images/base-app/scripts/wait-x11:1]: Executable

File: images/pegasus/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        wget -O cemu-emu.AppImage "${CEMU_APP_IMAGE_URL}"
+
+# Pegasus still requires libssl1.1, but ubuntu switched to libssl3. Need to manually install it.
+# TODO: Doing this could be a terrible idea (it does work)
+RUN wget http://archive.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.1_1.1.1f-1ubuntu2_amd64.deb --output-document=libssl1.1.deb && \
+    dpkg -i libssl1.1.deb
+
+# Install pegasus
+# ARG PEGASUS_DOWNLOAD=https://github.com/mmatyas/pegasus-frontend/releases/download/weekly_2023w20/pegasus-frontend_0.16.51_amd64.deb
+ARG PEGASUS_DOWNLOAD=https://github.com/mmatyas/pegasus-frontend/releases/download/continuous/pegasus-frontend_0.16.75_amd64.deb
```
[the remark this answers is not in the document]
Comment (dafrenchyman) [reply images/pegasus/Dockerfile:115]: I was reading that there's a new version of pegasus where they fixed this odd dependency on the old `libssl`1.1. They updated it to `libssl` 3, so maybe it's time to update it. However, I see what you mean about manually building it. They have a `continuous` release with a changing `.deb` file in there.

[the remark this answers is not in the document]
Comment (dafrenchyman) [reply images/pegasus/Dockerfile:115]: We can always revert to the one I commented out above. That one doesn't change. I'm also asking them in their discord channel if they could potentially make a newer release.

[the remark this answers is not in the document]
Comment (dafrenchyman) [reply images/pegasus/Dockerfile]: I fixed the above!

File: images/pegasus/Dockerfile [hunk 5]
```diff
@@ -55,103 +52,70 @@ RUN \
...
+
+  # Pegasus still requires libssl1.1, but ubuntu switched to libssl3. Need to manually install it.
+  # TODO: Doing this could be a terrible idea (it does work)
+  wget http://archive.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.1_1.1.1f-1ubuntu2_amd64.deb --output-document=libssl1.1.deb
+  apt-get install -y ./libssl1.1.deb
+
+  # Pegasus doesn't tag releases, they have a rolling weekly release
+  # so we have to parse the Github APIs to get the link
+  curl -s https://api.github.com/repos/mmatyas/pegasus-frontend/releases/tags/continuous | \
+  jq -r '.assets[] | select(.name | contains(".deb")) | .browser_download_url' | \
```
Comment (dafrenchyman) [inline images/pegasus/Dockerfile:72]: OMG - You already did it!

Comment (dafrenchyman) [reply images/pegasus/Dockerfile:72]: If you're getting the latest version from continuous, we "may" be able to skip the `libssl1.1.deb` bit. I was in their discord channel a while back and they mentioned that they had updated the code to use the newer `libssl3`.

**Comment by ABeltramo** [comment]
## Recap

We've got `sway` as an alternative to `gamescope` for non-fullscreen apps in the `base-app` image. You can switch the compositor running on startup with the following env variables:
 - `RUN_GAMESCOPE=1`: start Gamescope
 - `RUN_SWAY=1`: start Sway
 - if you don't pass anything it'll just `exec ${APP_TO_LAUNCH}`

## How to use

```bash
docker build -t gow/base images/base
docker build -t gow/base-app --build-arg BASE_IMAGE=gow/base images/base-app
docker build -t gow/pegasus --build-arg BASE_APP_IMAGE=gow/base-app images/pegasus
```

Then in Wolf `config.toml`
```toml
[[apps]]
title = "Emulators"
start_virtual_compositor = true

[apps.runner]
type = "docker"
name = "WolfEmulators"
image = "gow/pegasus"
mounts = []
env = [
"RUN_SWAY=1",
"GOW_REQUIRED_DEVICES=/dev/input/event* /dev/dri/* /dev/nvidia*",
]
devices = []
ports = []
base_create_json = """
{
  "HostConfig": {
    "IpcMode": "host",
    "CapAdd": ["NET_RAW", "MKNOD", "NET_ADMIN", "SYS_ADMIN", "SYS_NICE"],
    "Privileged": false,
    "DeviceCgroupRules": ["c 13:* rmw"]
  }
}
\
"""
```

## What's still not working?

It seems that I can't properly click things when using Sway. I think that it might be something specific with QT for the following reasons:
 - Running the Firefox container in Sway works perfectly
 - I can see the warning message: `qt.qpa.plugin: Could not find the Qt platform plugin "wayland" in ""`
   - I've tried installing `qtwayland5` and `qt6-wayland` following [this](https://unix.stackexchange.com/questions/598099/could-not-find-the-qt-platform-plugin-wayland) but that doesn't seem to work. I also wonder if it's related to <https://github.com/swaywm/sway/issues/4857>.

## TODO
- [ ] merge es-de and retroarch into a single dockerfile
- [ ] investigate joypad support for emulators
- [ ] see if we can drop ~~all~~ most of the hardcoded config files
- [ ] properly expose and map paths so that everything is under `~/retro`

**Pull request merged**